src/query/functions/utils/heap.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package utils
+
+import (
+	"container/heap"
+)
+
+// ValueIndexPair is a pair of float value and index at which it exists
+type ValueIndexPair struct {
+	Val   float64
+	Index int
+}
+
+type lessFn func(ValueIndexPair, ValueIndexPair) bool
+
+func maxHeapLess(i, j ValueIndexPair) bool {
+	if i.Val == j.Val {
+		return i.Index > j.Index
+	}
+	return i.Val < j.Val
+}
+
+func minHeapLess(i, j ValueIndexPair) bool {
+	if i.Val == j.Val {
+		return i.Index > j.Index
+	}
+	return i.Val > j.Val
+}
+
+// FloatHeap is a heap that can be given a maximum size
+type FloatHeap struct {
+	isMaxHeap bool
+	capacity  int
+	floatHeap *floatHeap
+}
+
+// NewFloatHeap builds a new FloatHeap based on first parameter
+// and a capacity given by second parameter. Zero and negative
+// values for maxSize provide an unbounded FloatHeap
+func NewFloatHeap(isMaxHeap bool, capacity int) FloatHeap {
+	var less lessFn
+	if isMaxHeap {
+		less = maxHeapLess
+	} else {
+		less = minHeapLess
+	}
+
+	floatHeap := &floatHeap{
+		heap: make([]ValueIndexPair, 0, capacity),
+		less: less,
+	}
+
+	heap.Init(floatHeap)
+	return FloatHeap{
+		isMaxHeap: isMaxHeap,
+		capacity:  capacity,
+		floatHeap: floatHeap,
+	}
+}
+
+// Push pushes a value and index pair to the heap
+func (fh FloatHeap) Push(value float64, index int) {
```

Comment:
don't think this is the right way. Why not push the element to the heap and keep removing until we are <= capacity ? I think you're trying to get into the heap internals a lot which we should avoid here.

Reply:
also my other earlier comment about just heapifying the whole thing instead of individual pushes, might change this a bit.

Reply:
This is much better in the scenario we have 10,000 series and we're taking top 20, much better to have a max heap size to cap it to 20 elements, saving both memory and time needed to sort the heap on each insert/pop

Reply:
You're right about this being optimal but this requires us to be more careful about correctness. I don't think a slice of 10K vs a slice of 20 really matters that much since we can reuse the same slice anyway. My suggestion was to keep the code simpler by not making this optimization until its really needed. 
If you will still like to go forward with your approach, that's fine but we'll then have to take a closer look at your logic...

Reply:
That's the point of the review isn't it 😛